near_api/providers.py

```diff
@@ -18,28 +23,104 @@ class FinalityTypes:
 
 
 class JsonProviderError(Exception):
-    pass
+    def get_type(self) -> Optional[str]:
+        try:
+            return self.args[0]["name"]
+        except (IndexError, KeyError):
+            return None
+
+    def get_cause(self) -> Optional[str]:
+        try:
+            return self.args[0]["cause"]["name"]
+        except (IndexError, KeyError):
+            return None
+
+    def is_invalid_nonce_tx_error(self) -> bool:
+        try:
+            return (
+                self.get_type() == "HANDLER_ERROR"
+                and self.get_cause() == "INVALID_TRANSACTION"
+                and "InvalidNonce" in self.args[0]["data"]["TxExecutionError"]["InvalidTxError"]
+            )
+        except (IndexError, KeyError):
+            return False
+
+    def is_expired_tx_error(self) -> bool:
+        try:
+            return (
+                self.get_type() == "HANDLER_ERROR"
+                and self.get_cause() == "INVALID_TRANSACTION"
+                and self.args[0]["data"]["TxExecutionError"]["InvalidTxError"] == "Expired"
+            )
+        except (IndexError, KeyError):
+            return False
 
 
 class JsonProvider(object):
-    def __init__(self, rpc_addr, proxies=None):
+    def __init__(
+            self,
+            rpc_addr,
+            proxies=None,
+            tx_timeout_retry_number: int = 12,
+            tx_timeout_retry_backoff_factor: float = 1.5,
+            http_retry_number: int = 10,
+            http_retry_backoff_factor: float = 1.5,
+            session: Optional[requests.Session] = None,
```

Comment:
Is this argument ever used? I’d rather we get rid of it.  With it, the prototype of this constructor is a bit weird.  If session is given, majority of the arguments are just ignored.

Reply:
Hm... since you prefer to not use built-in retry mechanism then it's not very useful at the moment. I'll remove it from this PR.

---

The main reason for adding this here was to make it possible to configure HTTP connection pooling(*) which is very useful when doing more than couple of requests to near api per second.

I intended to do something like:
```
session = requests.Session()
sessions.mount(
  "https://",
  requests.HTTPAdapter(
    pool_maxsize=1,
    pool_connections=10,
    pool_block=True,
    ...
  )
)

provider = near_api.providers.JsonProvider(..., session=session)
```

I'll remove this argument here and create a separate PR which will allow me to do:
```
provider = near_api.providers.JsonProvider(..., http_pool_maxsize=1, http_pool_block=...)
```

under the hood it will create session with relevant pool size and `retries=0`. So the client won't be staring a new TLS connection each time and client will be handling retires itself. Sounds ok?

(*) relevant docs: https://requests.readthedocs.io/en/latest/api/#requests.adapters.HTTPAdapter

Reply:
I mean it’s fine to have optional session argument.  My issue was about having mutually exclusive set of arguments.